### ISSUE TITLE
[prefer-readonly-type] False positive when  allowMutableReturnType is enabled.

### ISSUE THREAD

**Issue by RebeccaStevens** [jonaskello/eslint-plugin-functional#98]
The follow will emit a linting error even if the option `allowMutableReturnType` is enabled.

```ts
function foo(bar: string): { baz: number } {
  //                       ~~~~~~~~~~~~~~~ A readonly modifier is required.
}
```

**Comment by peterreisz** [comment]
Linting type definition files (attached to js based libs) also gives you false positive with this option turned on:

```ts
export function adjust<T>(index: number, replaceFn: (x: T) => T, list: readonly T[]): T[];
export function adjust<T>(index: number, replaceFn: (x: T) => T): (list: readonly T[]) => T[];
```

```
  1:87  error  Only readonly arrays allowed  functional/prefer-readonly-type
  2:91  error  Only readonly arrays allowed  functional/prefer-readonly-type
 ```

https://github.com/selfrefactor/rambda/issues/565
